src/relax/training/utils.cc

```diff
@@ -0,0 +1,177 @@
+  // TODO(chaofan): a better way to check whether new_ret_var should be dataflow
+  void RemapToDataflow(SeqExpr body) {
+    for (const BindingBlock& block : body->blocks) {
+      for (const Binding& binding : block->bindings) {
+        const auto* binding_node = binding.as<VarBindingNode>();
+        if (binding_node && !binding_node->var->IsInstance<DataflowVarNode>()) {
+          Var new_binding_var = DataflowVar(
+              binding_node->var->vid, GetStructInfo(binding_node->var), binding_node->var->span);
+          this->var_remap_[binding_node->var->vid] = new_binding_var;
+        }
+      }
+    }
+  }
+
+  Array<Var> RemapLossParams(const Array<Var>& loss_func_params, Array<Var> new_params) {
+    for (int i = 0; i < static_cast<int>(loss_func_params.size()); ++i) {
+      Var loss_param = loss_func_params[i];
+      auto loss_param_sinfo = GetStructInfo(loss_param);
+
+      if (i < static_cast<int>(orig_rets.size())) {  // map return value to loss param
+        auto orig_ret_sinfo = GetStructInfo(orig_rets[i]);
+        ICHECK(StructuralEqual()(orig_ret_sinfo, loss_param_sinfo))
+            << "The struct info of the " << i
+            << "-th return value of orig func is: " << orig_ret_sinfo
+            << " while the corresponding struct info of parameter of loss function is "
+            << loss_param_sinfo << ", which is different.";
+
+        if (const auto* var_node = orig_rets[i].as<VarNode>()) {
+          ICHECK(orig_rets[i].as<DataflowVarNode>());
+          orig_rets_var_.push_back(NullOpt);
+          this->var_remap_[loss_param->vid] = GetRef<Var>(var_node);
+        } else {
+          Var new_ret_var = DataflowVar(/*name_hint=*/"ret_" + std::to_string(i), orig_ret_sinfo);
+          orig_rets_var_.push_back(new_ret_var);
+          this->var_remap_[loss_param->vid] = new_ret_var;
+        }
+      } else {  // append to the param list
+        Var new_loss_param = Var(loss_param->vid, loss_param_sinfo, loss_param->span);
+        this->var_remap_[loss_param->vid] = new_loss_param;
+        new_params.push_back(new_loss_param);
+      }
+    }
+    return new_params;
+  }
+
+  Array<Expr> orig_rets;
+
+ private:
+  SeqExpr loss_body_;
+  Array<Optional<Var>> orig_rets_var_;
+};
+
+/*!
+ * \brief Local helper to append a specified loss function after the original function.
+ * \param orig_func The function to be appended to.
+ * \param loss_func The loss function.
+ * \return The result function after appended.
+ */
+Function AppendLoss(Function orig_func, Function loss_func) {
+  CHECK(orig_func->body->IsInstance<SeqExprNode>())
+      << "The body of the original function is expected to be a SeqExpr, but got"
+      << orig_func->body->GetTypeKey();
+  CHECK(loss_func->body->IsInstance<SeqExprNode>())
+      << "The body of the loss function is expected to be a SeqExpr, but got"
+      << loss_func->body->GetTypeKey();
+
+  auto param_copied_func = CopyWithNewParams(orig_func);
```

Comment:
Nit: Try to avoid using `auto` when the type does not explicitly appear in the statement.